test/unit/uri.c

```diff
@@ -583,16 +590,195 @@ test_invalid_string_uri_set(void)
+
+const struct str_escape escape_testcase[] = {
+	[0] = {
+		.str = "-._~",
+		.escaped = "-._~",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[1] = {
+		.str = "0123456789",
+		.escaped = "0123456789",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[2] = {
+		.str = "abcdefghijklm",
+		.escaped = "abcdefghijklm",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[3] = {
+		.str = "nopqrstuvwxyz",
+		.escaped = "nopqrstuvwxyz",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[4] = {
+		.str = "ABCDEFGHIJKLM",
+		.escaped = "ABCDEFGHIJKLM",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[5] = {
+		.str = "NOPQRSTUVWXYZ",
+		.escaped = "NOPQRSTUVWXYZ",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[6] = {
+		.str = "!$&'()*+,;=",
+		.escaped = "%21%24%26%27%28%29%2A%2B%2C%3B%3D",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+};
+
+/**
+ * Builds an array with unreserved characters.
+ * uri.unreserved() implemented as a Lua function,
+ * unreserved_tbl() replaces Lua implementation for testing purposes.
+ */
+static void
+unreserved_tbl(const char *str, unsigned char unreserved[256])
+{
+	for (int i = 0; i < 256; i++)
+		unreserved[i] = 0;
+
+	for (; *str; str++) {
+		unsigned char ch = (unsigned char)*str;
+		unreserved[ch] = 1;
+	}
+}
+
+static void
+test_escape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	unsigned char unreserved[256];
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *unescaped = escape_testcase[i].str;
+		const char *escaped = escape_testcase[i].escaped;
+		bool plus = escape_testcase[i].plus;
+		char *dst = xcalloc(strlen(unescaped) * 3, sizeof("char"));
+		unreserved_tbl(escape_testcase[i].unreserved, unreserved);
+		size_t dst_size = uri_escape(unescaped, strlen(unescaped),
+					     dst, unreserved, plus);
+		isnt(dst_size, 0, "escaped string ('%s') length != 0", dst);
+		is(strcmp(escaped, dst), 0, "escape: '%s' == '%s'",
+		   escaped, dst);
+		free(dst);
+	}
+	check_plan();
+	footer();
+}
+
+static void
+test_unescape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *unescaped = escape_testcase[i].str;
+		const char *escaped = escape_testcase[i].escaped;
+		bool decode_plus = escape_testcase[i].plus;
+		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
+		size_t dst_size = uri_unescape(escaped, strlen(escaped),
+					       dst, decode_plus);
+		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
+		     dst);
+		is(strcmp(dst, unescaped), 0, "unescape: '%s' == '%s'",
+		   unescaped, dst);
+		free(dst);
+	}
+	check_plan();
+	footer();
+}
+
+const struct str_escape unescape_testcase[] = {
+	/* Special case: %<non-hex><non-hex> */
+	[0] = {
+		.str = "%##",
+		.escaped = "%##",
+		.unreserved = "%%#",
+		.plus = false,
+	},
```

Comment:
`unreserved` is not used in those test cases (because they're about decoding). Moreover, if we would call `uri_escape()` with such `unreserved` parameter, it would lead to access out of buffer bounds: the `unreserved` array much be `const unsgned char[256]`, i.e. must contain 256 bytes.

I propose to set `NULL` to `unreserved` for those test cases.

Reply:
Ouch, I see, it is an array of characters to transform into 256 bytes array later. That's correct. So I don't insist on `NULL` here.

(Side note: we don't need to lua-pattern-escape `%` here. IOW, no need to duplicate `%`. However it is harmless.)

I'll close this discussion.